### ISSUE TITLE
🐛 Bug Report: Multiple failures in monorepo with parent yarn workspaces

### ISSUE THREAD

**Issue by ourchitectureio** [backstage/backstage#23161]
### 📜 Description

The default Backstage application as well as the `backstage-cli` commands seem to make several assumptions about the file system that may not be true e.g. a monorepo with Yarn setup in a parent directory.

### 👍 Expected behavior

Built on `yarn`, Backstage shouldn't care if it's the root or a child.

### 👎 Actual Behavior with Screenshots

Notice that the "Reproduction steps" has issues like:

- `yarn build:all` will initiate the `backstage-cli` and will search for "yarn.lock" in the Backstage directory rather than in the monorepo root.
- Jest may completely fail with errors surrounding "rootDir" under a "./src" directory within the Backstage project that doesn't exist because of [this line](https://github.com/backstage/backstage/blob/4e627d45938fbe41c77ca65de6e8bb0a465b4c9c/packages/cli/config/jest.js#L130)
- Trying to provide a custom Jest configuration completely ignores the Backstage Jest configuration
- Trying to extend the Backstage configuration by importing it and expanding it into an extended config map ignorantly will cause an infinite loop because of [this code](https://github.com/backstage/backstage/blob/4e627d45938fbe41c77ca65de6e8bb0a465b4c9c/packages/cli/config/jest.js#L70-L77) e.g. custom jest config would load backstage jest config which would load custom jest config which would... ♾️
- Copying all of the Backstage Jest configuration into the project makes progress, but with deeper and deeper issues like complaining Playwright tests, `window is not defined` issues, and more.
- ... probably more than this 😁

### 👟 Reproduction steps

Example: create a new monorepo, initialize a new yarn project and upgrade it to the latest, stable version with `yarn set version stable`, add a dependency example to simulate a larger monorepo like `yarn add -D prettier` with a task in "package.json" by adding `"scripts": { "lint": "prettier --check ." } }`, configure future workspace support in "package.json" by adding `"workspaces": [ "src/**/*" ]`, create a subdirectory like `mkdir --parents ./src/examples/ && cd ./src/examples/`, and then create the Backstage code base with the command `npx @backstage/create-app@latest`. Finally, remove the "yarn.lock" file since it is in the root of the monorepo.

## No known workaround(s)

Unfortunately, [`yarn` workspaces does not allow an "ignore" pattern](https://github.com/yarnpkg/yarn/issues/7051). As a result, it seemed that the only option is to modify the workspace pattern matching not to match the Backstage directory path; possibly moving the Backstage project to a path outside of the parent yarn workspaces entirely. However, something about the tooling still makes wrong assumptions. For example, the `yarn test` command in the root of the Backstage project fails with "Validation Error: Directory X:\src\github\ourchitecture\monorepo\src\systems\dev\backstage\ourstage\src in the rootDir option was not found." And, the command `yarn test -- --rootDir ./` processes for a long time and fails with two out of three failing tests; one fails because it attempted to run the Playwright test outside of the `yarn playwright test` command and the other fails with "windows is not defined" and an environment error "The error below may be caused by using the wrong test environment, see https://jestjs.io/docs/configuration#testenvironment-string. Consider using the "jsdom" test environment.".

### 🖥️ Your Environment

$ yarn backstage-cli info
yarn run v1.22.21
$ X:\src\somestage\node_modules\.bin\backstage-cli info
OS:   Windows_NT 10.0.22631 - win32/x64
node: v20.11.1
yarn: 1.22.21
cli:  0.25.2 (installed)
backstage:  1.23.0

Dependencies:
  @backstage/app-defaults                                          1.5.0
  @backstage/backend-app-api                                       0.5.11
  @backstage/backend-common                                        0.21.0
  @backstage/backend-dev-utils                                     0.1.4
  @backstage/backend-openapi-utils                                 0.1.3
  @backstage/backend-plugin-api                                    0.6.10
  @backstage/backend-tasks                                         0.5.15
  @backstage/catalog-client                                        1.6.0
  @backstage/catalog-model                                         1.4.4
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.3
  @backstage/cli                                                   0.25.2
  @backstage/config-loader                                         1.6.2
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.12.0
  @backstage/core-compat-api                                       0.2.0
  @backstage/core-components                                       0.14.0
  @backstage/core-plugin-api                                       1.9.0
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.5
  @backstage/frontend-plugin-api                                   0.6.0
  @backstage/integration-aws-node                                  0.1.9
  @backstage/integration-react                                     1.1.24
  @backstage/integration                                           1.9.0
  @backstage/plugin-api-docs                                       0.11.0
  @backstage/plugin-app-backend                                    0.3.58
  @backstage/plugin-app-node                                       0.1.10
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.2
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.1
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.5
  @backstage/plugin-auth-backend-module-github-provider            0.1.7
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.7
  @backstage/plugin-auth-backend-module-google-provider            0.1.7
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.5
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.7
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.3
  @backstage/plugin-auth-backend-module-oidc-provider              0.1.0
  @backstage/plugin-auth-backend-module-okta-provider              0.0.3
  @backstage/plugin-auth-backend                                   0.21.0
  @backstage/plugin-auth-node                                      0.4.5
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.7
  @backstage/plugin-catalog-backend                                1.17.0
  @backstage/plugin-catalog-common                                 1.0.21
  @backstage/plugin-catalog-graph                                  0.4.0
  @backstage/plugin-catalog-import                                 0.10.6
  @backstage/plugin-catalog-node                                   1.7.0
  @backstage/plugin-catalog-react                                  1.10.0
  @backstage/plugin-catalog                                        1.17.0
  @backstage/plugin-events-node                                    0.2.19
  @backstage/plugin-github-actions                                 0.6.11
  @backstage/plugin-org                                            0.6.20
  @backstage/plugin-permission-common                              0.7.12
  @backstage/plugin-permission-node                                0.7.21
  @backstage/plugin-permission-react                               0.4.20
  @backstage/plugin-proxy-backend                                  0.4.8
  @backstage/plugin-scaffolder-backend-module-azure                0.1.2
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.2
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.0
  @backstage/plugin-scaffolder-backend-module-github               0.2.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.13
  @backstage/plugin-scaffolder-backend                             1.21.0
  @backstage/plugin-scaffolder-common                              1.5.0
  @backstage/plugin-scaffolder-node                                0.3.0
  @backstage/plugin-scaffolder-react                               1.8.0
  @backstage/plugin-scaffolder                                     1.18.0
  @backstage/plugin-search-backend-module-catalog                  0.1.14
  @backstage/plugin-search-backend-module-pg                       0.5.19
  @backstage/plugin-search-backend-module-techdocs                 0.1.14
  @backstage/plugin-search-backend-node                            1.2.14
  @backstage/plugin-search-backend                                 1.5.0
  @backstage/plugin-search-common                                  1.2.10
  @backstage/plugin-search-react                                   1.7.6
  @backstage/plugin-search                                         1.4.6
  @backstage/plugin-tech-radar                                     0.6.13
  @backstage/plugin-techdocs-backend                               1.9.3
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.5
  @backstage/plugin-techdocs-node                                  1.11.2
  @backstage/plugin-techdocs-react                                 1.1.16
  @backstage/plugin-techdocs                                       1.10.0
  @backstage/plugin-user-settings                                  0.8.1
  @backstage/release-manifests                                     0.0.11
  @backstage/repo-tools                                            0.6.0
  @backstage/test-utils                                            1.5.0
  @backstage/theme                                                 0.5.1
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by tudi2d** [comment]
Hey. Can you maybe provide a bit more information on your setup & where exactly Backstage fit's in there? Why is a standalone repository not a solution for you? 

What looks a bit confusing to me is the "remove yarn.lock" part. You probably should not remove the `yarn.lock` file from the Backstage instance. 

If I understand your `yarn` setup correctly you are looking to "nest" a workspace in another workspace, which is currently not supported (by design): https://github.com/yarnpkg/yarn/pull/6151. This is a decission by `yarn` on their tool & nothing we should hack around in the cli.

**Comment by ourchitectureio** [comment]
Hey @tudi2d, yes there are a few things:

Not trying to "nest" workspaces. Instead, we're trying to define workspaces at a higher level and `yarn` works perfectly fine with this. Backstage tooling seems to make assumptions that it "has" (owns?) the "yarn stuff" (e.g. workspaces and "yarn.lock"); which it does not need to assume.

1. I feel it's a safe design assumption to make that Backstage could exist as a sub-directory in a larger monorepo that contains "other things"
2. I feel it's a safe design assumption that `yarn` might be used in a parent directory to Backstage and that, since `yarn workspaces` can work at any level, Backstage should work "just fine" with the "yarn.lock" in a parent directory and as one set of Backstage workspaces (app, backend, plugins) in a larger ecosystem of other project workspaces e.g. a CLI, other services, other apps, etc.

Given those assumptions, in the example provided above, "yarn.lock" exists in a parent directory to Backstage. So, it isn't "removed". Yarn supports this and simply doesn't care. Yarn even picks up the Backstage workspaces perfectly.

However, the Backstage CLI and related tooling seems to:

- Assume that Backstage is the owner of the yarn workspaces ecosystem. Yarn doesn't make this a requirement. In theory, Backstage tooling should respect yarn being run from a parent directory (one or more levels up) with an assumption that "yarn.lock" will exist in that parent directory.
- Assume that Backstage is at the root of where `yarn` commands are being run from. For example, the `yarn test:all` works when Backstage is installed stand-alone, but fails when `yarn` is installed in a parent directory with workspaces that include the Backstage workspaces.

**Comment by tudi2d** [comment]
Thanks for the longer explantions. We do see the repository tooling, such as the `backstage-cli`, as utilities for the most common use cases. It should make adopting the developer portal, as well as work on this repository, easier. You always got the chance to extend the utilities yourself for a more specific use case that goes outside of what we are aiming for by forking those packages.

What is still a bit unclear for me right now is why you are choosing this setup? What are you trying to build & what is Backstage in this context? Maybe I'm missing a common use case, but it sounds to me like you are trying to build something new on top of Backstage (which is totally fine - but ouside of our scope to support).

**Comment by danoc** [comment]
Hi @tudi2d, I ran into this issue recently at Thumbtack and can speak to it a bit more.

> What is still a bit unclear for me right now is why you are choosing this setup? What are you trying to build & what is Backstage in this context? Maybe I'm missing a common use case, but it sounds to me like you are trying to build something new on top of Backstage (which is totally fine - but ouside of our scope to support).

A few years ago we began consolidating all of our frontend services into a Yarn monorepo. As part of this effort we've created a handful of useful internal libraries that our services use:

– Automatic telemetry with OpenTelemetry (loaded with Node.js's `--require` flag)
– Structured JSON logging with Pino (loaded with Node.js's `--require` flag)
– Configs for tooling like linters, Prettier, etc.
– Scripts to inject secrets into our Node.js processes

We've also set up task running and caching with Turborepo so that the monorepo runs quickly.

We were hoping to include Backstage as yet another web/Node.js based service within our monorepo so that it can benefit from the internal infrastructure/tooling that we've built and so that our product engineers can contribute plugins in a codebase that they're already familiar with.

> If I understand your yarn setup correctly you are looking to "nest" a workspace in another workspace, which is currently not supported (by design): https://github.com/yarnpkg/yarn/pull/6151. This is a decission by yarn on their tool & nothing we should hack around in the cli.

This isn't the case in Yarn 2+ (Berry). It's possible to configure the`<root-dir>/package.json` file such that `<root-dir>/backstage/package.json` is a workspace and`<root-dir>/backstage/packages/<package>/package.json` are also also workspaces.

---

As @ourchitectureio pointed out, the main issue is that the Backstage CLIs expect that Backstage's `package.json` will always be in the same directory as `yarn.lock`. Maybe this work could be bundled in with the work to support other package managers like pnpm (https://github.com/backstage/backstage/issues/20658)? It seems that @aramissennyeydd [has a similar issue](https://github.com/backstage/backstage/issues/20658#issuecomment-2203185292) and that @Rugvip has already [made this problem easier to solve](https://github.com/backstage/backstage/pull/25714).
